website/docs/tutorials/react-rsc.md

```diff
@@ -0,0 +1,198 @@
+};
+
+export default function RootLayout({ params: { lang }, children }: Props) {
+  const i18n = allI18nInstances[lang]; // get a ready-made i18n instance for the given locale
+  setI18n(i18n); // make it available server-side for the current request
+
+  return (
+    <html lang={lang}>
+      <body>
+        <LinguiClientProvider initialLocale={lang} initialMessages={i18n.messages}>
+          <YourApp />
+        </LinguiClientProvider>
+      </body>
+    </html>
+  );
+}
+```
+
+Step (2) is implemented in `LinguiClientProvider`, which is a client component:
+
+```tsx title="LinguiClientProvider.tsx"
+"use client";
+
+import { I18nProvider } from "@lingui/react";
+import { type Messages, setupI18n } from "@lingui/core";
+import { useState } from "react";
+
+export function LinguiClientProvider({
+  children,
+  initialLocale,
+  initialMessages,
+}: {
+  children: React.ReactNode;
+  initialLocale: string;
+  initialMessages: Messages;
+}) {
+  const [i18n] = useState(() => {
+    return setupI18n({
+      locale: initialLocale,
+      messages: { [initialLocale]: initialMessages },
+    });
+  });
+  return <I18nProvider i18n={i18n}>{children}</I18nProvider>;
+}
+```
+
+:::tip
+Why we are not passing the I18n instance directly from `RootLayout` to the client via `LinguiClientProvider`? It's because the I18n object isn't serializable, and cannot be passed from server to client.
+:::
+
+Lastly, there's the `appRouterI18n.ts` file, which is only executed on server and holds one instance of I18n object for each locale of our application. See [here](#) how it's implemented in the example.
+
+### Rendering translations in server and client components
+
+Below you can see an example of a React component. This component can be rendered **both with RSC and on client**. This is great if you're migrating a Lingui-based project from pages router to app router because you can keep the same components working in both worlds.
+
+In fact, if you swapped the html tags for their more universal alternatives, this component could also be used in React Native.
+
+```tsx title="app/[lang]/components/SomeComponent.tsx"
+import { Trans, t } from "@lingui/macro";
+
+export function SomeComponent() {
+  const { i18n } = useLingui();
+  return (
+    <div>
+      <p>
+        <Trans>Some Item</Trans>
+      </p>
+      <p>{t(i18n)`Other Item`}</p>
+    </div>
+  );
+}
+```
+
+As you might recall, hooks are not supported in RSC, so you might be surprised that this works. Under RSC, `useLingui` is actually not a hook but a simple function call which reads from the React `cache` mentioned above.
+
+Here's how you could implement server-side `useLingui` yourself using `getI18n`, which is another way to obtain the I18n instance on the server:
+
+```tsx title="useLinguiForRSC.ts"
+import { getI18n } from "@lingui/react/server";
+
+export function useLingui() {
+  const ctx = getI18n();
+  if (!ctx) {
+    // throw an informative error
+  }
+
+  return ctx;
+}
+```
```

Comment:
do we really need this block? I believe the purpose of it is just informational, but that may bring a bit of cluttering and confusion for the reader. So developer may think that it need to implement this function by itself.  

I think the previous paragraph is totally enough to describe the idea behind RSC version of useLingui

Reply:
agree. I removed it, but linked to the implementation. I think it's good to give people understanding of how things work inside.